### PR TITLE
extended gridedit api to allow invoking cell editor by row/column

### PR DESCRIPTION
**Pull request by justin-romano**
new to pull requests, hope im doing it right.
I have not put much here  because im hoping the documentation on the methods is mostly enough.
My main concern is around security. Im mostly used to working in Dojo which does not hide much and was hoping that adding an array of column scopes column directive to the row object was in fitting with the way things are done here.
```
            if (!$scope.row.cols) {
              $scope.row.cols = [];
            }
            $scope.row.cols[$scope.colRenderIndex] = { "$scope": $scope, "$elm": $elm };
```

### CODE REVIEW

**Comment by justin-romano** [comment]
So what happens now?

File: src/features/edit/js/gridEdit.js [hunk 1]
```diff
@@ -112,12 +112,45 @@
               }
             },
             methods: {
-              edit: { }
+              edit: {
+                /**
+                 * @ngDoc method
+                 * @name editCell
```
Comment (JLLeitschuh) [inline src/features/edit/js/gridEdit.js]: You need a @methodOf param here.
Example:
https://github.com/angular-ui/ui-grid/blob/master/src/features/cellnav/js/cellnav.js#L418

File: src/features/edit/js/gridEdit.js [hunk 2]
```diff
@@ -112,12 +112,45 @@
...
+                 */
+                editCell: function (row, col, persist) {
+                  if (persist) {
+                    this.rows[row].cols[col].$scope.persistEdit = true;
+                  }
+                  this.rows[row].cols[col].$scope.$broadcast('editCell');
+                },
+                /**
+                 * @ngDoc method
+                 * @name editEditCell
```
Comment (JLLeitschuh) [inline src/features/edit/js/gridEdit.js]: You need a @methodOf param here.
Example:
https://github.com/angular-ui/ui-grid/blob/master/src/features/cellnav/js/cellnav.js#L418

File: src/features/edit/js/gridEdit.js [hunk 3]
```diff
@@ -112,12 +112,45 @@
               }
             },
             methods: {
-              edit: { }
+              edit: {
+                /**
+                 * @ngDoc method
```
Comment (JLLeitschuh) [inline src/features/edit/js/gridEdit.js]: This should be function not method

File: src/features/edit/js/gridEdit.js [hunk 4]
```diff
@@ -112,12 +112,45 @@
...
+                 * @param {boolean} persist the cell in edit mode. Prevent cancel events such as scroll from removing the editor
+                 */
+                editCell: function (row, col, persist) {
+                  if (persist) {
+                    this.rows[row].cols[col].$scope.persistEdit = true;
+                  }
+                  this.rows[row].cols[col].$scope.$broadcast('editCell');
+                },
+                /**
+                 * @ngDoc method
```
Comment (JLLeitschuh) [inline src/features/edit/js/gridEdit.js]: This needs to be `function` not `method`

**Comment by JLLeitschuh** [comment]
If it's not too much to ask it would be really great if you could add a unit and/or e2e test for this.

**Comment by justin-romano** [comment]
I was going to wait to see if you guys would accept it before a put time
into that. Will do ASAP

File: src/features/edit/js/gridEdit.js (hunk 3 above)
Comment (justin-romano) [inline src/features/edit/js/gridEdit.js]: Sorry nube to this framework. so in the context of this framework what is the definition of a method?
In my understanding a function is not a property of an object.

Comment (justin-romano) [inline src/features/edit/js/gridEdit.js]: e.g.
```
/**
* @ngdoc function
* @name angular.merge
* @module ng
* @kind function
*
* @description
* Deeply extends the destination object `dst` by copying own enumerable properties from the `src` object(s)
* to `dst`. You can specify multiple `src` objects. If you want to preserve original objects, you can do so
* by passing an empty object as the target: `var object = angular.merge({}, object1, object2)`.
*
* Unlike {@link angular.extend extend()}, `merge()` recursively descends into object properties of source
* objects, performing a deep copy.
*
* @param {Object} dst Destination object.
* @param {...Object} src Source object(s).
* @returns {Object} Reference to `dst`.
*/
function merge(dst) {
  return baseExtend(dst, slice.call(arguments, 1), true);
}
```

Comment (justin-romano) [inline src/features/edit/js/gridEdit.js]: vs...
```javascript
 /**
         * @ngdoc method
         * @name $cookies#put
         *
         * @description
         * Sets a value for given cookie key
         *
         * @param {string} key Id for the `value`.
         * @param {string} value Raw value to be stored.
         * @param {Object=} options Options object.
         *    See {@link ngCookies.$cookiesProvider#defaults $cookiesProvider.defaults}
         */
        put: function(key, value, options) {
          $$cookieWriter(key, value, calcOptions(options));
        },
```

**Comment by swalters** [comment]
I don't think we would ever assign anything column related to the row. It feels very wrong to me.  The marriage of Rows and Columns is done through a nested repeat, so you have $scope.row and $scope.col on every cell directive.

What are you trying to accomplish?

**Comment by justin-romano** [comment]
Thank Shane. I know exactly what you mean. Feels dirty to me but given the code structure I could not easily think of a way to get the scope for a column so I could then apply it to the cell editor. For my project I want to invoke the editor of a cell so when I add a new row I can place a cell in edit mode for convenience of ui. Let me know if you can think of a better way

**Comment by justin-romano** [comment]
so how about this approach?
```javascript
                editCell: function (row, col, persist) {
                  var cellScope = this.element.find("div[id*=" + row.toString() + "-" + this.columns[col].uid + "]").scope();
                  if (persist) {
                    cellScope.persistEdit = true;
                  }
                  cellScope.$broadcast('editCell');
```

**Comment by JLLeitschuh** [comment]
Make sure the the element you are trying to grab isn't a directive that has `replace=true`. Don't know if this is at all related but you may just want to check that.

**Comment by justin-romano** [comment]
Thanks. Its set in the uiGrid service but only after rendering complete. I may need to mock it the tests if thats allowed

**Comment by JLLeitschuh** [comment]
Thats today in 2 hours

**Comment by swalters** [comment]
@justin-romano this looks much better to me.  You have some failing jshint's.  also, please rebase to squash the commits. thanks

**Comment by JLLeitschuh** [comment]
Looks good. Take care of @swalters requests and we'll get this merged up.

**Comment by swalters** [comment]
@justin-romano will this work when angular debug info is turned off?
https://docs.angularjs.org/guide/production

**Comment by mportuga** [comment]
Closing this due to inactivity.

**Pull request closed without merging**